numpy/ma/core.py

```diff
@@ -3893,77 +3894,45 @@ def _delegate_binop(self, other):
                 return True
         return False
 
-    def __eq__(self, other):
-        """
-        Check whether other equals self elementwise.
-
-        """
+    def _comparison(self, other, op):
         if self is masked:
             return masked
         omask = getattr(other, '_mask', nomask)
-        if omask is nomask:
-            check = self.filled(0).__eq__(other)
-            try:
-                check = check.view(type(self))
-                check._mask = self._mask
-            except AttributeError:
-                # Dang, we have a bool instead of an array: return the bool
-                return check
+        odata = other if omask is nomask else filled(other, 0)
+        check = getattr(self.filled(0), op)(odata)
+
+        if isinstance(check, (bool_, bool)):
+            return masked if self._mask else check
+
+        if self._mask is nomask:
+            mask = omask
         else:
-            odata = filled(other, 0)
-            check = self.filled(0).__eq__(odata).view(type(self))
-            if self._mask is nomask:
-                check._mask = omask
-            else:
-                mask = mask_or(self._mask, omask)
-                if mask.dtype.names:
-                    if mask.size > 1:
-                        axis = 1
-                    else:
-                        axis = None
-                    try:
-                        mask = mask.view((bool_, len(self.dtype))).all(axis)
-                    except ValueError:
-                        mask = np.all([[f[n].all() for n in mask.dtype.names]
-                                       for f in mask], axis=axis)
-                check._mask = mask
+            mask = mask_or(self._mask, omask)
+            if mask.dtype.names:
+                # Structured array.  Mask element if all parts are masked.
+                mask = mask.view(bool_).reshape(mask.shape+(-1,)).all(-1)
```

Comment:
Now that you ask, I have been going back and forth on it...

After some thought, the only consistent result I see is to return `masked` for all three cases, except that we need some extra code so that `m == m`, ie

```python
>>> p = np.ma.mvoid((1,7), mask=(0,1), dtype='i4,f4')
>>> (m == p), (m == m)
True, True
```
which I think requires replacement of `mask_or` by `mask_xor`, and then use `any` instead of `all`.

Returning `True` seems worse to me because intuitively I think `n` and `m` don't represent equal values, and secondly because the code would have to be much more complicated: We would have to reimplements numpy's internal structure comparison (somewhere in the C code) inside of MaskedArray to ignore fields with masked values.